rsconf/component/sirepo_sandbox.py

```diff
@@ -0,0 +1,125 @@
+:license: http://www.apache.org/licenses/LICENSE-2.0.html
+"""
+from pykern import pkio
+from pykern.pkcollections import PKDict
+from pykern.pkdebug import pkdp
+from rsconf import component
+import subprocess
+
+
+class T(component.T):
+    def install_resource(self, *args, **kwargs):
+        super().install_resource(*args, component_name="sirepo_sandbox", **kwargs)
+
+    def internal_build_compile(self):
+        from rsconf import systemd
+
+        if self.name == "sirepo_sandbox":
+            for u in self.hdb.sirepo_sandbox.users.keys():
+                t = T(u, self.buildt)
+                self.buildt.build_component(t)
+            return
+        self.buildt.require_component("docker", "network")
+        jc, _ = self.j2_ctx_init()
+        z = jc.sirepo_sandbox
+        z.setdefault("volumes", {})
+        # TODO(e-carlin): Better way then building the path manually
+        # /srv/<component-name> is output by systemd.docker_unit_prepare
+        # below but to call that we need z.host.ssh_d. z.host needs z.host_d
+        # so we are in a dependency cycle
+        z.host_d = z.host_root_d.join(self.name, "db")
+        z.secrets = self._gen_keys(jc, z)
+        for x in "guest", "host":
+            z[x] = self._gen_paths(z, z.get(x + "_d"))
+        z.run_u = jc.rsconf_db.run_u
+        # Only additional config for the server is the sshd config.
+        z.run_d = systemd.docker_unit_prepare(self, jc, watch_files=[z.host.ssh_d])
+        self._network(jc, z)
+
+    def internal_build_write(self):
+        from rsconf import systemd
+
+        if self.name == "sirepo_sandbox":
+            self.append_root_bash(": nothing to do")
+            return
+        jc = self.j2_ctx
+        z = jc.sirepo_sandbox
+        v = []
+        x = [
+            [z.host.ssh_d, z.guest.ssh_d],
+            # SECURITY: no modifications to run_d
+            [z.run_d, z.run_d, "ro"],
+        ]
+        for d in z.volumes:
+            h = z.host[d]
+            x.append([h, z.guest[d]])
+            v.append(h)
+        systemd.docker_unit_enable(
+            self,
+            jc,
+            image=jc.sirepo_sandbox.docker_image,
+            volumes=x,
+            cmd="/usr/sbin/sshd -D -f '{}'".format(z.guest.sshd_config),
+        )
+        self.install_access(mode="700", owner=z.run_u)
+        for d in [z.host_d, z.host.ssh_d, *v]:
+            self.install_directory(d)
+        self.install_access(mode="400")
+        self.install_resource(z.host.sshd_config, jc)
+        for k, v in z.secrets.items():
+            self.install_abspath(v, z.host[k])
+
+    def _network(self, jc, z):
+        n = self.buildt.get_component("network")
+        z.ip, _ = n.ip_and_net_for_host(jc.rsconf_db.host)
+        z.ssh_port = jc.sirepo_sandbox.users[self.name]
+
+    def _gen_keys(self, jc, z):
+        from rsconf import db
+
+        res = PKDict()
+        b = db.secret_path(
+            jc,
+            "sirepo_sandbox/" + self.name,
```

Comment:
I don't think this needs to be on the channel, it can be just on the host. There's no channel relationship that needs to be maintained. Fix secret_path to take a parameter "directory=True" in this case that would do a mkdir_parent instead of a mkdir_parent_only. This is probably used in other places.

Reply:
1f49f6a